### PR TITLE
Prompty is callable

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
- Prompty as flow
   - [x] Prompty callable: Flow.load(source="xxxx")(**inputs)
       - [x] support asyn
   - [x] Test/Batch run: pf.run(flow="xxx.prompty", data="datas/path")
   - [x] prompty in experiment
 - Prompty be referenced in flow
   - [x] Prompty referenced by eager flow

Next PR:
- [ ] support image input
- [ ] Add trace to prompty 
- [ ] Prompty referenced by DAG flow

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+from typing import Dict, Union
+
+from promptflow._constants import FlowLanguage
+from promptflow._sdk._constants import BASE_PATH_CONTEXT_KEY
+from promptflow.exceptions import ErrorTarget, UserErrorException
+
+from .base import FlowBase
+
+
+class Prompty(FlowBase):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow/prompty.py:15]: give this class some docstring

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        # prompty file path
+        path = Path(path)
+        # prompty folder path
+        code = Path(code)
+        super().__init__(code=code, path=path, data=data, content_hash=None, **kwargs)
+
+    # region overrides
+
+    @property
+    def language(self) -> str:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow/prompty.py:32]: why prompty has language?

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        return FlowLanguage.Python
+
+    @property
+    def additional_includes(self) -> list:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow/prompty.py:36]: does prompty has this?

File: src/promptflow/promptflow/core/__init__.py [hunk 4]
```diff
@@ -7,9 +7,9 @@
...
 # flake8: noqa
 from promptflow._core.tool import ToolProvider, tool
-from promptflow.core._flow import AsyncFlow, Flow
+from promptflow.core._flow import AsyncFlow, Flow, Prompty
 
 # backward compatibility
 log_flow_metric = log_metric
 
-__all__ = ["log_metric", "ToolProvider", "tool", "Flow", "AsyncFlow"]
+__all__ = ["log_metric", "ToolProvider", "tool", "Flow", "AsyncFlow", "Prompty"]
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/__init__.py:15]: Let's expose this when the feature fully developed

File: src/promptflow/promptflow/core/_flow.py [hunk 5]
```diff
@@ -52,12 +65,15 @@ def load(
...
-        with open(flow_path, "r", encoding=DEFAULT_ENCODING) as f:
-            flow_content = f.read()
-            data = load_yaml_string(flow_content)
-        if is_flex_flow(yaml_dict=data, working_dir=flow_dir):
-            raise UserErrorException("Please call entry directly for flex flow.")
-        return cls._create(code=flow_dir, path=flow_path, data=data)
+        if is_prompty_flow(flow_path):
+            return Prompty.load(source=flow_path, **kwargs)
+        else:
+            with open(flow_path, "r", encoding=DEFAULT_ENCODING) as f:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:71]: Do we have a function called load_yaml_file?

File: src/promptflow/promptflow/core/_flow.py [hunk 6]
```diff
@@ -167,3 +183,197 @@ async def invoke(self, inputs: dict, *, connections: dict = None, **kwargs) -> "
...
+    Prompty as function example:
+    .. code-block:: python
+
+        from promptflow.core import Prompty
+        prompty = Prompty.load(source="path/to/prompty.prompty")
+        result = prompty(input_a=1, input_b=2)
+
+    """
+
+    def __init__(self, path: Union[str, PathLike], **kwargs):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:221]: The init class should have args like: parameters, api, connection?

File: src/promptflow/promptflow/core/_flow.py [hunk 7]
```diff
@@ -167,3 +183,197 @@ async def invoke(self, inputs: dict, *, connections: dict = None, **kwargs) -> "
...
+        for input_name, value in self.inputs.items():
+            if input_name not in input_values and "default" not in value:
+                missing_inputs.append(input_name)
+                continue
+            resolved_inputs[input_name] = input_values.get(input_name, value.get("default", None))
+        if missing_inputs:
+            raise MissingRequiredInputError(f"Missing required inputs: {missing_inputs}")
+        return resolved_inputs
+
+    def __call__(self, *args, **kwargs):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:309]: what's current trace tree now? share an example?

File: src/promptflow/promptflow/core/_utils.py [hunk 8]
```diff
@@ -110,6 +121,269 @@ def generate_flow_meta(
...
 
+def render_jinja_template_content(template_content, *, trim_blocks=True, keep_trailing_newline=True, **kwargs):
+    template = Template(template_content, trim_blocks=trim_blocks, keep_trailing_newline=keep_trailing_newline)
+    return template.render(**kwargs)
+
+
+def get_connection(connection):
+    if not isinstance(connection, (str, dict)):
+        error_message = """
+        Illegal definition of connection. Need to provide connection name or connection info like below:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_utils.py:132]: shall we point user to some doc link online?

Comment (wangchao1230) [inline src/promptflow/promptflow/core/_utils.py:132]: maybe we just say we expect str and dict here?

File: src/promptflow/promptflow/core/_utils.py [hunk 9]
```diff
@@ -110,6 +121,269 @@ def generate_flow_meta(
...
+        f"Not Support connection type {connection_type} for embedding api. "
+        f"Connection type should be in [{AzureOpenAIConnection.TYPE}, {OpenAIConnection.TYPE}]."
+    )
+    raise InvalidConnectionTypeError(message=error_message)
+
+
+# region: Copied from promptflow-tools
+
+
+def normalize_connection_config(connection):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_utils.py:166]: shall we move new utils to core/_prompty_utils.py?

File: src/promptflow/promptflow/executor/_prompty_executor.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+from promptflow.contracts.tool import InputDefinition
+from promptflow.core._flow import Prompty
+from promptflow.storage import AbstractRunStorage
+from promptflow.tracing._trace import _traced
+
+from ._script_executor import ScriptExecutor
+
+
+class PromptyExecutor(ScriptExecutor):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/executor/_prompty_executor.py:12]: add some docstring describe the dfiferent with PromptyExecutor & ScriptExecutor

File: src/promptflow/promptflow/contracts/flow.py [hunk 11]
```diff
@@ -316,6 +317,16 @@ def deserialize(data: dict) -> "Node":
         :return: The node constructed from the dict.
         :rtype: ~promptflow.contracts.flow.Node
         """
+        from promptflow._utils.flow_utils import is_prompty_flow
+
+        if data.get("source", {}).get("path", None) and is_prompty_flow(data.get("source").get("path")):
```
[the remark this answers is not in the document]
Comment (lalala123123) [reply src/promptflow/promptflow/contracts/flow.py:322]: This is in the scenario that prompty is referencing in dag flow.
When the source file is prompty, the prompty config will be parsed and used as the default value for the field in node

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/entities/_flow/prompty.py:15]: add docstring to it

File: src/promptflow/promptflow/core/_flow.py (hunk 5 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/_flow.py:70]: There is no such utility function. This logic is also not suitable for extraction, and some calls need to be processed with content

File: src/promptflow/promptflow/executor/_prompty_executor.py (hunk 10 above)
Comment (lalala123123) [reply src/promptflow/promptflow/executor/_prompty_executor.py:12]: add docstring

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/entities/_flow/prompty.py:15]: add docstring

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/entities/_flow/prompty.py]: additional_include and language will be used in batch run and test. The call has been modified and the two properties have been removed

File: src/promptflow/promptflow/_sdk/entities/_flow/prompty.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/entities/_flow/prompty.py]: additional_include and language will be used in batch run and test. The call has been modified and the two properties have been removed

File: src/promptflow/promptflow/_sdk/schemas/_flow.py [hunk 12]
```diff
@@ -92,3 +92,12 @@ def validate_entry(self, data, **kwargs):
 
         if entry_regex is not None and not re.match(entry_regex, data["entry"]):
             raise ValidationError(field_name="entry", message=f"Entry function {data['entry']} is not valid.")
+
+
+class PromptySchema(BaseFlowSchema):
+    """Schema for prompty."""
+
+    api = fields.Str(required=True)
```
[the remark this answers is not in the document]
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/schemas/_flow.py]: add default value to api

File: src/promptflow/promptflow/core/__init__.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/__init__.py]: remove it

File: src/promptflow/promptflow/core/_utils.py (hunk 9 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/_utils.py]: move to _prompty_utils.py

File: src/promptflow/promptflow/core/_utils.py (hunk 8 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/_utils.py]: fix it

Comment (lalala123123) [reply src/promptflow/promptflow/core/_utils.py]: add link doc to connection creation

File: src/promptflow/promptflow/core/_flow.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/_flow.py]: add those params

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 13]
```diff
@@ -536,6 +543,8 @@ def _check_eager_flow_and_language_from_yaml(self):
         # TODO: remove this after path is removed
         if flow_file.suffix.lower() == ".dll":
             return True, FlowLanguage.CSharp
+        elif flow_file.suffix.lower() == PROMPTY_EXTENSION:
+            return True, FlowLanguage.Python
```
Comment (wangchao1230) [inline src/promptflow/promptflow/batch/_batch_engine.py:547]: Prompty_Exetension does not mean it's using python runtime?

File: src/promptflow-core/promptflow/core/_flow.py [hunk 14]
```diff
@@ -180,3 +195,232 @@ async def invoke(self, inputs: dict, *, connections: dict = None, **kwargs) -> "
...
+        from promptflow.core import Prompty
+        prompty = Prompty.load(source="path/to/prompty.prompty")
+        result = prompty(input_a=1, input_b=2)
+
+    """
+
+    def __init__(
+        self,
+        path: Union[str, PathLike],
+        api: str = None,
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_flow.py:236]: model config parameter?

File: src/promptflow/tests/test_configs/prompty/prompty_with_image_input.prompty [hunk 15]
```diff
@@ -0,0 +1,18 @@
+---
+name: Prompty with image
+description: A basic prompt with image input
+init:
```
Comment (wangchao1230) [inline src/promptflow/tests/test_configs/prompty/prompty_with_image_input.prompty:4]: model?

Comment (lalala123123) [reply src/promptflow/tests/test_configs/prompty/prompty_with_image_input.prompty]: fix it

File: src/promptflow/promptflow/core/_flow.py (hunk 7 above)
Comment (lalala123123) [reply src/promptflow/promptflow/core/_flow.py]: Improve trace in next PR

**Review by wangchao1230 — DISMISSED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**